File: Training.Demo.Code.2020.Object.Oriented.Programming/C++/Composition/Camera.cpp
#include<iostream>

using namespace std;
class Flash {
private:
	bool isFlashOn;

public:
	Flash() {
        isFlashOn = false;
    }

    void setFlashOn(bool flashOn) {
        isFlashOn = flashOn;
    }

     void flashLight() {
        if (isFlashOn) {
            cout<<"Flashing light."<<endl;
        }
    }
};

class Camera {
private:
	Flash flash;

public:
	void clickPicture() {
        flash.flashLight();
    }

    void setFlashOn(bool flashOn) {
        flash.setFlashOn(flashOn);
    }

    ~Camera() {
    	// Destruction of the camera causes the destruction of flash.
    	flash.~Flash();
    }
};

int main(void) {
    Camera camera;
    camera.setFlashOn(true);
    camera.clickPicture();
    camera.~Camera();
}

File: Java/Composition/Camera.java
public class Camera {
    private Flash flash;

    public Camera() {
        flash = new Flash();
    }

    public void clickPicture() {
        flash.flashLight();
    }

    public void setFlashOn(boolean flashOn) {
        flash.setFlashOn(flashOn);
    }

    private class Flash {
        private boolean isFlashOn;

        public Flash() {
            isFlashOn = false;
        }

        public void setFlashOn(boolean flashOn) {
            isFlashOn = flashOn;
        }

        public void flashLight() {
            if (isFlashOn) {
                System.out.println("Flashing light.");
            }
        }
    }
}
